src/stirling/source_connectors/socket_tracer/protocols/common/data_stream_buffer.h

```diff
@@ -92,10 +92,23 @@ class DataStreamBuffer {
 
   /**
    * Get timestamp recorded for the data at the specified position.
+   * If less than previous timestamp, timestamp will be adjusted to be monotonically increasing.
    * @param pos The logical position of the data.
    * @return The timestamp or error if the position does not contain valid data.
    */
-  StatusOr<uint64_t> GetTimestamp(size_t pos) const { return impl_->GetTimestamp(pos); }
+  StatusOr<uint64_t> GetTimestamp(size_t pos) {
+    StatusOr<uint64_t> timestamp_ns_status = impl_->GetTimestamp(pos);
+    if (!timestamp_ns_status.ok()) return timestamp_ns_status;
+    uint64_t current_timestamp_ns = timestamp_ns_status.ConsumeValueOrDie();
+    uint64_t prev_timestamp_ns = GetPrevTimestamp();
+    if (current_timestamp_ns < prev_timestamp_ns) {
+      LOG(WARNING) << "Detected non-monotonically increasing timestamp " << current_timestamp_ns
+                   << ". Adjusting to previous timestamp + 1: " << prev_timestamp_ns + 1;
+      current_timestamp_ns = prev_timestamp_ns + 1;
+    }
+    SetPrevTimestamp(current_timestamp_ns);
```

Comment:
IIUC we do not need the setter/getter pattern for the previous timestamp member. Would recommend just directly accessing the private member here.